File: Split_Linked_List_in_Parts.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        # step 1
        length = 0
        cur = head
        while cur:
            length += 1
            cur = cur.next
        
        # step 2
        part_size = length // k
        larger_parts = length % k

        res = []

        # step 3
        cur = head
        for i in range(k):
            sublist_size = part_size + 1 if i < larger_parts else part_size
            if sublist_size == 0:
                res.append(None)
            else:
                sublist_head = cur
                for _ in range(sublist_size - 1):
                    cur = cur.next
                
                next_node = cur.next
                cur.next = None
                res.append(sublist_head)
                cur = next_node
        
        return res

File: Split_Linked_List_in_Parts.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        // Step 1: Calculate the length of the linked list
        int length = 0;
        ListNode* current = head;
        while (current) {
            length += 1;
            current = current->next;
        }

        // Step 2: Determine the size of each part
        int partSize = length / k;
        int largerParts = length % k;

        // Initialize a vector to store the results
        vector<ListNode*> result(k);

        // Step 3: Split the linked list into parts
        current = head;
        for (int i = 0; i < k; i++) {
            int sublistSize = (i < largerParts) ? (partSize + 1) : partSize;
            if (sublistSize == 0) {
                result[i] = nullptr;
            } else {
                ListNode* sublistHead = current;
                for (int j = 0; j < sublistSize - 1; j++) {
                    current = current->next;
                }
                ListNode* nextNode = current->next;
                current->next = nullptr;
                result[i] = sublistHead;
                current = nextNode;
            }
        }

        return result;        
    }
};
